Map/Map.cpp

```diff
@@ -0,0 +1,358 @@
+Graph& Graph::operator=(const Graph& rhs){
+    if(this != &rhs){
+        //create a deep copy of V
+        for(auto i=0;i<rhs.getV().size();i++){
+            V.push_back(new Node(*rhs.V[i]));
+        }
+        
+        //create a deep copy of listOfContinents
+        for(int i=0;i<rhs.getListOfContinents().size();i++){
+            listOfContinents.push_back(new Continent(*rhs.listOfContinents[i]));
+        }
+    }
+    return *this;
+}
+
+//-------------- Getters --------------//
+vector<Node*> Graph::getV() const{
+    return this->V;
+}
+
+vector<Continent*> Graph::getListOfContinents() const {
+    return this->listOfContinents;
+}
+
+//-------------- Inserting and connecting territories --------------//
+void Graph::insertATerritory(Territory data){
+    Node* newTerritory = new Node(data);
+    V.push_back(newTerritory);
+}
+
+void Graph::insertAndConnectTwoTerritories(Territory dataA, Territory dataB){
+    //1- create two territories without connections
+    insertATerritory(dataA);
+    insertATerritory(dataB);
+    //2- make the connection between the two territories
+    connectTwoNodes(getV().end()[-1],getV().end()[-2]);
+}
+
+void Graph::connectTwoNodes(Node* A, Node* B){
+    string edgeName = A->getData().getTerritoryName() + B->getData().getTerritoryName(); //AB
+    //cout<<edgeName<<endl;
+    A->addEdge(edgeName);
+    B->addEdge(edgeName);
+}
+
+//-------------- Others --------------//
+bool Graph::areConnected(Node* A, Node* B){
+    string possibleEdge1 = A->getData().getTerritoryName() + B->getData().getTerritoryName(); //AB
+    string possibleEdge2 = B->getData().getTerritoryName() + A->getData().getTerritoryName(); //BA
+    
+    for(string edge : A->getE()){//loop through all of the edges of A
+        if(edge == possibleEdge1 || edge == possibleEdge2)
+            return true;
+         }
+
+    return false;
+}
+
+void Graph::validate(){
+    bool isErrorThrown = false;
+
+    //check that the map is a connected graph
+    try{
+        if(!DFS(getV(),getV()[0]))
+            throw std::logic_error("Unconnected Graph");
+
+    }catch(const std::exception& e){
+        isErrorThrown = true;
+        std::cerr<<e.what()<<endl;
+    }
+
+    //check if all continents are connected sub-graphs
+    try{
+        for(Continent* continent : getListOfContinents()){
+            vector<Node*> tempListOfTerritoriesInContinent;
+            //getting all territories with the same continent in a single vector
+            for(Node* node : getV()){
+                if(node->getData().getContinent() == continent)
+                    tempListOfTerritoriesInContinent.push_back(node);
+            }
+            
+            if(!DFS(tempListOfTerritoriesInContinent,tempListOfTerritoriesInContinent[0]))
+                throw std::logic_error("Unconnected Sub-Graph: "+ continent->getContinentName() + " continent is not connected.");
+
+            tempListOfTerritoriesInContinent.clear();//earse the vector contentss
```

Comment:
I think this is a [memory leak](https://stackoverflow.com/questions/12795196/clearing-a-vector-of-pointers). Because clearing the vector doesn’t delete the pointers. So the pointers need to be deleted before the vector is cleared.
for(Node* node : tempListOfTerritoriesInContinent){
    delete[] node;
}
tempListOfTerritoriesInContinent.clear();

Reply:
Good point, but here i'm not creating pointers. I'm just storing existing ones in an array. These pointers are being destroyed later in the destructor.